File: app/src/main/java/com/yashovardhan99/healersdiary/Activities/PatientView.java
package com.yashovardhan99.healersdiary.Activities;

import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.yashovardhan99.healersdiary.Helpers.HtmlCompat;
import com.yashovardhan99.healersdiary.R;

import java.text.NumberFormat;
import java.util.Map;
import java.util.Objects;

import javax.annotation.Nullable;

public class PatientView extends AppCompatActivity {

    public String Uid;
    public final int REQUEST_HEALING_CONFIRMATION = 0;
    public final int REQUEST_PAYMENT_ADDED = 1;
    public final int REQUEST_FEEDBACK_ADDED = 2;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_patient_view);

        setSupportActionBar((Toolbar) findViewById(R.id.toolbar));
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setTitle("Patient Detail");
        //toolbar setup

        Uid = getIntent().getStringExtra(MainActivity.PATIENT_UID);
        //patient record UID to fetch records from firestore
        if (Uid != null) {

            final FirebaseFirestore db = FirebaseFirestore.getInstance();
            //initialize firestore
            final DocumentReference documentReference = db.collection(MainActivity.USERS)
                    .document(Objects.requireNonNull(FirebaseAuth.getInstance().getUid()))
                    .collection("patients")
                    .document(Uid);
            //create ref to patient document
            documentReference.addSnapshotListener(new EventListener<DocumentSnapshot>() {
                //listener to update patient data
                @Override
                public void onEvent(@Nullable DocumentSnapshot document, @Nullable FirebaseFirestoreException e) {
                    if (document!=null && document.exists()) {
                        //get document data
                        Log.d(MainActivity.FIRESTORE, "Data exists : " + document.getData());
                        //store it in a map and update relevant fields
                        Map<String, Object> patient = document.getData();//data extracted

                        if (patient == null)
                            return;

                        TextView name = findViewById(R.id.patientNameInDetail);
                        if (patient.containsKey("Name"))
                            name.setText(patient.get("Name").toString());

                        TextView disease = findViewById(R.id.patientDiseaseInDetail);
                        if (patient.containsKey("Disease") && !patient.get("Disease").toString().isEmpty())
                            disease.setText(patient.get("Disease").toString());
                        else
                            disease.setVisibility(View.GONE);

                        TextView due = findViewById(R.id.bill);
                        if (patient.containsKey("Due") && !patient.get("Due").toString().isEmpty()) {
                            Double amt = Double.parseDouble(patient.get("Due").toString());
                            String famt = getString(R.string.payment_due)+": <b><big>"+NumberFormat.getCurrencyInstance().format(amt)+"</big></b>";
                            due.setText(HtmlCompat.fromHtml(famt));
                        }
                        else
                            due.setVisibility(View.GONE);
                    }
                }

            });
        }

        //new healing button - to add new healing
        Button newHealing = findViewById(R.id.newHealingButton);
        newHealing.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent newHealingIntent = new Intent(PatientView.this,NewHealingRecord.class);
                newHealingIntent.putExtra(MainActivity.PATIENT_UID,Uid);
                startActivityForResult(newHealingIntent, REQUEST_HEALING_CONFIRMATION);
            }
        });

        //go to healing logs
        Button healingLogs = findViewById(R.id.healingLogsButton);
        healingLogs.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent logs = new Intent(PatientView.this, HealingLogs.class);
                logs.putExtra("PATIENT_NAME",((TextView)findViewById(R.id.patientNameInDetail)).getText());
                logs.putExtra(MainActivity.PATIENT_UID,Uid);
                startActivity(logs);
            }
        });

        //add new payment
        Button addPayment = findViewById(R.id.enterNewPPayment);
        addPayment.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent pay = new Intent(PatientView.this,PatientAddPaymentDialog.class);
                pay.putExtra(MainActivity.PATIENT_UID, Uid);
                startActivityForResult(pay,REQUEST_PAYMENT_ADDED);
            }
        });

        //go to payment logs
        Button paymentLogs = findViewById(R.id.viewPatientPaymentLogs);
        paymentLogs.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent logs = new Intent(PatientView.this, PatientPaymentLogs.class);
                logs.putExtra(MainActivity.PATIENT_UID,Uid);
                startActivity(logs);
            }
        });

        //add new patient feedback
        Button newFeedback = findViewById(R.id.newFeedbackButton);
        newFeedback.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent newFeed = new Intent(PatientView.this, NewPatientFeedback.class);
                newFeed.putExtra(MainActivity.PATIENT_UID, Uid);
                startActivityForResult(newFeed,REQUEST_FEEDBACK_ADDED);
            }
        });
    }

    @Override
    public void onBackPressed() {
        Intent back = new Intent(this,MainActivity.class);
        back.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        startActivity(back);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode){
            case REQUEST_HEALING_CONFIRMATION:
                if(resultCode==NewHealingRecord.NEW_HEALING_ADDED_RESULT)
                    Snackbar.make(findViewById(R.id.patientNameInDetail), R.string.added, Snackbar.LENGTH_SHORT).show();
                break;

            case REQUEST_PAYMENT_ADDED:
                if(resultCode==PatientAddPaymentDialog.PAYMENT_ADDED_RESULT)
                    Snackbar.make(findViewById(R.id.patientNameInDetail),getString(R.string.payment_added ,data.getStringExtra("Amount")),Snackbar.LENGTH_LONG).show();
                break;

            case REQUEST_FEEDBACK_ADDED:
                if(resultCode==RESULT_OK)
                    Snackbar.make(findViewById(R.id.patientNameInDetail),R.string.added, Snackbar.LENGTH_SHORT).show();
                break;


            default: super.onActivityResult(requestCode, resultCode, data);
        }
    }
}

File: app/src/main/java/com/yashovardhan99/healersdiary/activities/PatientView.kt
package com.yashovardhan99.healersdiary.activities

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.EventListener
import com.google.firebase.firestore.FirebaseFirestore
import com.yashovardhan99.healersdiary.R
import com.yashovardhan99.healersdiary.fragments.GeneratePatientBill
import com.yashovardhan99.healersdiary.fragments.NewHealingDialog
import com.yashovardhan99.healersdiary.fragments.PatientAddPaymentDialog
import com.yashovardhan99.healersdiary.helpers.HtmlCompat
import java.text.NumberFormat

class PatientView : AppCompatActivity(), View.OnClickListener {

    private lateinit var uid: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_patient_view)

        setSupportActionBar(findViewById<View>(R.id.toolbar) as Toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.title = "Patient Detail"
        //toolbar setup

        uid = intent.getStringExtra(MainActivity.PATIENT_UID)
        //patient record UID to fetch records from firestore

        val db = FirebaseFirestore.getInstance()
        //initialize firestore
        val documentReference = db.collection(MainActivity.USERS)
                .document(FirebaseAuth.getInstance().uid!!)
                .collection("patients")
                .document(uid)
        //create ref to patient document
        documentReference.addSnapshotListener(EventListener { document, _ ->
            //listener to update patient data
            if (document != null && document.exists()) {
                //get document data
                Log.d(MainActivity.FIRESTORE, "Data exists : " + document.data!!)
                //store it in a map and update relevant fields
                val patient = document.data ?: return@EventListener//data extracted

                val name = findViewById<TextView>(R.id.patientNameInDetail)
                if (patient.containsKey("Name"))
                    name.text = patient["Name"]!!.toString()

                val disease = findViewById<TextView>(R.id.patientDiseaseInDetail)
                if (patient.containsKey("Disease") && patient["Disease"]!!.toString().isNotEmpty())
                    disease.text = patient["Disease"]!!.toString()
                else
                    disease.visibility = View.GONE

                val due = findViewById<TextView>(R.id.bill)
                if (patient.containsKey("Due") && patient["Due"]!!.toString().isNotEmpty()) {
                    val amt = java.lang.Double.parseDouble(patient["Due"]!!.toString())
                    val famt = getString(R.string.payment_due) + ": <b><big>" + NumberFormat.getCurrencyInstance().format(amt) + "</big></b>"
                    due.text = HtmlCompat.fromHtml(famt)
                } else
                    due.visibility = View.GONE
            }
        })

        //new healing button - to add new healing
        val newHealing = findViewById<Button>(R.id.newHealingButton)
        newHealing.setOnClickListener(this)

        //go to healing logs
        val healingLogs = findViewById<Button>(R.id.healingLogsButton)
        healingLogs.setOnClickListener {
            val logs = Intent(this@PatientView, HealingLogs::class.java)
            logs.putExtra("PATIENT_NAME", (findViewById<View>(R.id.patientNameInDetail) as TextView).text)
            logs.putExtra(MainActivity.PATIENT_UID, uid)
            startActivity(logs)
        }

        //add new payment
        val addPaymentButton = findViewById<Button>(R.id.enterNewPPayment)
        addPaymentButton.setOnClickListener {
            val addPayment = PatientAddPaymentDialog()
            addPayment.show(supportFragmentManager, "NewPayment")
        }

        //go to payment logs
        val paymentLogs = findViewById<Button>(R.id.viewPatientPaymentLogs)
        paymentLogs.setOnClickListener {
            val logs = Intent(this@PatientView, PatientPaymentLogs::class.java)
            logs.putExtra(MainActivity.PATIENT_UID, uid)
            startActivity(logs)
        }

        //add new patient feedback
        val newFeedback = findViewById<Button>(R.id.newFeedbackButton)
        newFeedback.setOnClickListener {
            val newFeed = Intent(this@PatientView, NewPatientFeedback::class.java)
            newFeed.putExtra(MainActivity.PATIENT_UID, uid)
            startActivityForResult(newFeed, REQUEST_FEEDBACK_ADDED)
        }

        //go to feedback logs
        val feedbackLogs = findViewById<Button>(R.id.patientFeedbackLogs)
        feedbackLogs.setOnClickListener {
            val logs = Intent(this@PatientView, PatientFeedbackLogs::class.java)
            logs.putExtra(MainActivity.PATIENT_UID, uid)
            startActivity(logs)
        }

        val bill = findViewById<Button>(R.id.generateBill)
        bill.setOnClickListener(this)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {

            REQUEST_FEEDBACK_ADDED -> if (resultCode == Activity.RESULT_OK)
                Snackbar.make(findViewById(R.id.patientNameInDetail), R.string.added, Snackbar.LENGTH_SHORT).show()
            else -> super.onActivityResult(requestCode, resultCode, data)
        }
    }

    fun paymentAdded(amt: String) {
        Snackbar.make(findViewById(R.id.patientNameInDetail), getString(R.string.payment_added, amt), Snackbar.LENGTH_LONG).show()
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.newHealingButton -> {
                val newHealing = NewHealingDialog()
                newHealing.show(supportFragmentManager, "HEALINGNEW")
            }
            R.id.generateBill -> {
                val generateBill = GeneratePatientBill()
                val bundle = Bundle()
                bundle.putString(MainActivity.PATIENT_UID, uid)
                generateBill.arguments = bundle
                generateBill.show(supportFragmentManager, "generatebill")
            }
        }
    }

    companion object {
        private const val REQUEST_FEEDBACK_ADDED = 2
    }
}
